x-pack/plugins/xpack_main/server/lib/telemetry/monitoring/get_beats_stats.js

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { get } from 'lodash';
+import { createQuery } from './create_query';
+
+const HITS_SIZE = 10000; // maximum hits to receive from ES with each search
+
+/*
+ * Create a set of result objects where each is the result of searching hits from Elasticsearch with a size of HITS_SIZE each time.
+ * @param {Object} server The server instance
+ * @param {function} callCluster The callWithRequest or callWithInternalUser handler
+ * @param {Array} clusterUuids The string Cluster UUIDs to fetch details for
+ * @param {Date} start Start time to limit the stats
+ * @param {Date} end End time to limit the stats
+ * @param {Number} page selection of hits to fetch from ES
+ * @param {Array} resultSet previously accumulated resultSet to combine with the next
+ * @param {Array} resultSet previously accumulated resultSet to combine with the next
+ * @return {Array} Array of 1 or more result objects
+ */
+export async function fetchBeatsStats(server, callCluster, clusterUuids, start, end, { page = 0, resultSet = [] } = {}) {
+  const config = server.config();
+
+  const params = {
+    index: config.get('xpack.monitoring.beats.index_pattern'),
+    ignoreUnavailable: true,
+    filterPath: [
+      'hits.hits._source.cluster_uuid',
+      'hits.hits._source.beats_stats.beat.version',
+      'hits.hits._source.beats_stats.beat.type',
+      'hits.hits._source.beats_stats.beat.host',
+      'hits.hits._source.beats_stats.metrics.libbeat.pipeline.events.published',
+      'hits.hits._source.beats_stats.metrics.libbeat.output.type',
+    ],
+    body: {
+      query: createQuery({
+        start,
+        end,
+        type: 'beats_stats',
+        filters: [{ terms: { cluster_uuid: clusterUuids } }],
+      }),
+      collapse: { field: 'beats_stats.beat.uuid', },
+      sort: [{ 'beats_stats.timestamp': 'desc' }],
+      from: page * HITS_SIZE,
+      size: HITS_SIZE
+    },
+  };
+
+  const results = await callCluster('search', params);
+
+  const hitsLength = get(results, 'hits.hits.length', -1);
+  resultSet.push(results);
```

Comment:
We should pre-process each payload before saving it to minimize the amount of data that we maintain in the heap during this process. Otherwise we will have every unique instance sitting in heap until the end.

Perhaps `handleBeatsResultSet` could be augmented to have the `clusters` object passed in so that it can just pickup where it left off while also only expecting a _single_ result set instead of the paged array?

Reply:
Sounds good. Are you thinking we pass in results from a single query as well as the `clusters` object to `handleBeatsResultSet`, which mutates `clusters` to have new processed stat data? The object mutation wouldn't worry me since the object's lifecycle exists just in this lib file.

Reply:
Exactly!

Reply:
4ba349319d